### PR TITLE
Command .indexes displaying indexes of database or a table.

### LINKED ISSUE #100: Add .indexes command
List all indexes:
```
genji> .indexes
```

List indexes of a specific table:
```
genji> .indexes user
```

### PR DESCRIPTION
**Pull request by tzzed**
This PR fixes #100. Two options for the command `.Indexes` and `.Indexes <table name>` are implemented. Tests `.tables` and `.indexes` are added. 

```
genji> .indexes
main skill index on users(skills.0)
idx_nen on users(nen)
genji>
```

Refactoring of  `struct Index`. Let me know what you think about it

### CODE REVIEW

File: cmd/genji/shell/shell.go [hunk 1]
```diff
@@ -189,7 +189,7 @@ func (sh *Shell) executeInput(in string) error {
 	switch {
 	// if it starts with a "." it's a command
 	// it must not be in the middle of a multi line query though
-	case strings.HasPrefix(in, ".") && sh.query == "":
+	case strings.HasPrefix(in, "."):
```
Comment (yaziine) [inline cmd/genji/shell/shell.go:192]: I am not sure to understand why did you remove the `&& sh.query == ""` statement. Could you elaborate please?

File: cmd/genji/shell/shell.go [hunk 2]
```diff
@@ -213,16 +213,29 @@ func (sh *Shell) executeInput(in string) error {
 	return nil
 }
 
-func (sh *Shell) runCommand(cmd string) error {
-	switch cmd {
+func (sh *Shell) runCommand(in string) error {
+	cmd := strings.Fields(in)
```
Comment (yaziine) [inline cmd/genji/shell/shell.go:217]: TIL! I used to use `strings.Split` for this kind of action. Thanks!

File: cmd/genji/shell/shell.go [hunk 3]
```diff
@@ -213,16 +213,29 @@ func (sh *Shell) executeInput(in string) error {
...
 			return err
 		}
-		return runTablesCmd(db)
+
+		return runTablesCmd(db, cmd)
 	case ".exit":
+		if len(cmd) > 1 {
+			return fmt.Errorf("too many arguments in call to %s", cmd[0])
+		}
+
```
Comment (yaziine) [inline cmd/genji/shell/shell.go:230]: I think that how it was done before was enough. Could you tell me what motivates you to change this? There is probably a use case that I don't see.

File: database/config.go [hunk 4]
```diff
@@ -298,10 +298,7 @@ func (i *IndexConfig) ScanDocument(d document.Document) error {
 type Index struct {
 	index.Index
 
-	IndexName string
-	TableName string
-	Path      document.ValuePath
-	Unique    bool
+	Opts IndexConfig
```
Comment (yaziine) [inline database/config.go:301]: First of all thanks for doing this! 

However, I suggest to embed directly `IndexConfig`:
```go
type Index struct {
    index.Index
    IndexConfig
}
```
like that instead of accessing the config fields through `Opts`
```go
idx.Opts.Path.GetValue(d)
```
we can directly do
```go
idx.Path.GetValue(d)
````

What do you think?

File: database/table.go [hunk 5]
```diff
@@ -554,3 +551,17 @@ func (t *Table) Indexes() (map[string]Index, error) {
 
 	return indexes, nil
 }
+
+// PrintIndexes prints a list of indexes of the table receiver.
+func (t *Table) PrintIndexes() error {
```
Comment (yaziine) [inline database/table.go:556]: Printing the indexes is a feature that is really linked to the shell. Thus, I suggest to move this logic to its package. 
I think that it's useful to have this function only there. What do you think?

File: cmd/genji/shell/command_test.go [hunk 6]
```diff
@@ -0,0 +1,82 @@
+package shell
+
+import (
+	"github.com/genjidb/genji"
+	"github.com/stretchr/testify/require"
+	"strings"
+	"testing"
+)
+
+func Test_runTablesCmd(t *testing.T) {
```
Comment (yaziine) [inline cmd/genji/shell/command_test.go:10]: I think we should have the same way of naming function tests within the codebase. Does it make sense?
```suggestion
func TestTablesCmd(t *testing.T) {
```

File: cmd/genji/shell/command_test.go [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+			defer db.Close()
+
+			if err := runTableIndexCmd(db, test.in); (err != nil) != test.wantErr {
+				require.Errorf(t, err, "", test.wantErr)
+			}
+		})
+	}
+}
+
+func Test_runIndexesCmd(t *testing.T) {
```
Comment (yaziine) [inline cmd/genji/shell/command_test.go:40]: Same as above
```suggestion
func TestIndexesCmd(t *testing.T) {
```

File: cmd/genji/shell/command.go [hunk 8]
```diff
@@ -24,3 +29,62 @@ func runTablesCmd(db *genji.DB) error {
...
+		//
+		if err == database.ErrTableNotFound {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// runIndexesCmd executes all indexes of the database or all indexes of the given table.
```
Comment (yaziine) [inline cmd/genji/shell/command.go:78]:
```suggestion
// runIndexesCmd executes the .indexes command.
```

File: cmd/genji/shell/command.go [hunk 9]
```diff
@@ -24,3 +29,62 @@ func runTablesCmd(db *genji.DB) error {
...
+func runIndexesCmd(db *genji.DB, in []string) error {
+	switch len(in) {
+	case 1:
+		// If the input is ".Indexes"
```
Comment (yaziine) [inline cmd/genji/shell/command.go:82]:
```suggestion
		// If the input is ".indexes"
```

File: cmd/genji/shell/command.go [hunk 10]
```diff
@@ -24,3 +29,62 @@ func runTablesCmd(db *genji.DB) error {
...
+		return runAllIndexesCmd(db)
+	case 2:
+		// If the input is ".Indexes <tableName>" cmd[1] is the table name
```
Comment (yaziine) [inline cmd/genji/shell/command.go:85]:
```suggestion
		// If the input is ".indexes <tableName>"
```

File: cmd/genji/shell/command.go [hunk 11]
```diff
@@ -24,3 +29,62 @@ func runTablesCmd(db *genji.DB) error {
...
+		return runTableIndexCmd(db, in[1])
+	}
+
+	return fmt.Errorf("too many arguments in call to %s", in[0])
```
Comment (yaziine) [inline cmd/genji/shell/command.go:89]:
```suggestion
	return fmt.Errorf("too many arguments to the command %s", in[0])
```

File: database/config.go (hunk 4 above)
Comment (yaziine) [reply database/config.go:301]: Actually, I am sure its a good idea because it will expose all method of the `IndexConfig`. What do you think @asdine ?

File: cmd/genji/shell/shell.go (hunk 1 above)
Comment (tzzed) [reply cmd/genji/shell/shell.go:192]: I did that because `.indexes` could take an argument `.indexes tableName` that led me to change the behaviour of`runCommand` by checking the len of the splitted input string. 
`sh.query == ""` was a 'guard' against input command with arguments but now we need it for `.indexes`.

File: cmd/genji/shell/shell.go (hunk 3 above)
Comment (tzzed) [reply cmd/genji/shell/shell.go:230]: See two comments above.

File: database/table.go (hunk 5 above)
Comment (tzzed) [reply database/table.go:556]: I got it.

File: database/config.go (hunk 4 above)
Comment (tzzed) [reply database/config.go:301]: Very good thing this avoids the refactoring of the majority of files.

Comment (asdine) [reply database/config.go:301]: Yeah I would keep it like this, embedding would make the `Index` type API weird

File: cmd/genji/shell/command.go [hunk 12]
```diff
@@ -24,3 +29,62 @@ func runTablesCmd(db *genji.DB) error {
...
 	return nil
 }
+
+// runTableIndexCmd prints the all indexes that the given table contains.
+func runTableIndexCmd(db *genji.DB, tableName string) error {
+	err := db.ViewTable(tableName, func(tx *genji.Tx, table *database.Table) error {
+		return table.PrintIndexes()
+	})
+
+	if err == document.ErrFieldNotFound {
```
[the remark this answers is not in the document]
Comment (tzzed) [reply cmd/genji/shell/command.go:39]: I tried to catch it before but no success, I should catch it before and did not break the error handler for another command or request.Need more investigation...

[the remark this answers is not in the document]
Comment (asdine) [reply cmd/genji/shell/command.go:39]: If you need help don't hesitate asking 👍

File: cmd/genji/shell/command_test.go [hunk 13]
```diff
@@ -0,0 +1,82 @@
+package shell
+
+import (
+	"github.com/genjidb/genji"
```
[the remark this answers is not in the document]
Comment (tzzed) [reply cmd/genji/shell/command_test.go:4]: now yes!

File: sql/query/insert.go [hunk 14]
```diff
@@ -54,6 +55,7 @@ func (stmt InsertStmt) insertDocuments(t *database.Table, stack expr.EvalStack)
 	var res Result
 
 	for _, e := range stmt.Values {
+		fmt.Println("values row:\n", e)
```
Comment (asdine) [inline sql/query/insert.go:58]: You forgot to remove this debug line 😉

File: cmd/genji/shell/command.go (hunk 8 above)
Comment (yaziine) [reply cmd/genji/shell/command.go:73]: @tzzed this comment has been marked as resolved but the suggestion is not applied. Probably due to an omission. Do you think you can apply it?

File: cmd/genji/shell/shell.go [hunk 15]
```diff
@@ -213,16 +213,29 @@ func (sh *Shell) executeInput(in string) error {
...
 		db, err := sh.getDB()
 		if err != nil {
 			return err
 		}
-		return runTablesCmd(db)
+
+		return runTablesCmd(db, cmd)
 	case ".exit":
+		if len(cmd) > 1 {
+			return fmt.Errorf("too many arguments in call to %s", cmd[0])
```
[the remark this answers is not in the document]
Comment (yaziine) [reply cmd/genji/shell/shell.go:229]: Same here @tzzed
